benchmark/src/pre_benchmarks.cpp

```diff
@@ -0,0 +1,319 @@
+        {NULL, 0, NULL, 0}
+    };
+
+    const char* optstring = "m:d:h";
+    while ((opt = getopt_long(argc, argv, optstring, long_options, NULL)) != -1) {
+        std::cerr << "opt1: " << opt << "; optarg: " << optarg << std::endl;
+        switch (opt) {
+        case 'm':
+            security_model = atoi(optarg);
+            break;
+        case 'd':
+            num_of_hops = atoi(optarg);
+            break;
+        case 'h':
+            usage();
+        default:
+            return false;
+        }
+    }
+
+    //0 - CPA secure PRE, 1 - fixed 20 bits noise, 2 - provable secure HRA noise flooding
+    parameters.SetPREMode(INDCPA);
+    parameters.SetKeySwitchTechnique(BV);
+     if (security_model == 0) {
+       plaintextModulus = 2;
+       multDepth = 0;
+       sigma = 3.2;		
+       securityLevel = HEStd_128_classic;
+       ringDimension = 1024;
+       relinWindow = 3;
+       dcrtbits = 0;
+
+       qmodulus = 27;
+       firstqmod = 27;
+       security_model = 0;
+       parameters.SetPREMode(INDCPA);
+    } else if (security_model == 1) {
+        plaintextModulus = 2;
+        multDepth = 0;
+        sigma = 3.2;		
+        securityLevel = HEStd_128_classic;
+        ringDimension = 2048;
+        relinWindow = 18;
+        dcrtbits = 0;
+
+        qmodulus = 54;
+        firstqmod = 54;
+        security_model = 1; 
+        parameters.SetPREMode(FIXED_NOISE_HRA);
+    } else if (security_model == 2) {
+        plaintextModulus = 2;
+        multDepth = 0;
+        sigma = 3.2;		
+        securityLevel = HEStd_128_classic;
+        ringDimension = 16384;
+        relinWindow = 1;
+        dcrtbits = 30;
+
+        qmodulus = 438;
+        firstqmod = 60;
+        security_model = 2;
+        parameters.SetPREMode(NOISE_FLOODING_HRA);
+    } else if (security_model == 3) {
+        ringDimension = 16384;
+        relinWindow = 0;
+        dcrtbits = 30;
+
+        qmodulus = 438;
+        firstqmod = 60;
+        uint32_t dnum = 3;
+        parameters.SetPREMode(NOISE_FLOODING_HRA);
+        parameters.SetKeySwitchTechnique(HYBRID);
+        parameters.SetNumLargeDigits(dnum);
+    }
+
+    parameters.SetMultiplicativeDepth(multDepth);
+    parameters.SetPlaintextModulus(plaintextModulus);
+	parameters.SetSecurityLevel(securityLevel);
+	parameters.SetStandardDeviation(sigma);
+	parameters.SetMaxDepth(2);
+    parameters.SetMode(OPTIMIZED);
+    parameters.SetRingDim(ringDimension);
+    parameters.SetFirstModSize(firstqmod);
+    parameters.SetScalingFactorBits(dcrtbits);
+    parameters.SetRelinWindow(relinWindow);
+    parameters.SetRescalingTechnique(FIXEDMANUAL);
+    parameters.SetMultiHopQModulusLowerBound(qmodulus);
+
+    cryptoContext = GenCryptoContext(parameters);
+            
+        // Enable features that you wish to use
+    cryptoContext->Enable(PKE);
+    cryptoContext->Enable(KEYSWITCH);
+    cryptoContext->Enable(LEVELEDSHE);
+    cryptoContext->Enable(PRE);
+
+
+    std::cout << "p = "
```

Comment:
I can remove these, but I have these to display the parameters for which the benchmark would run. I think it would be helpful when it runs. If we can leave it there, it would be good but if we have a policy of not having couts in benchmarking files, I can remove it.

Reply:
the production code must not print anything as this is not a research branch. it can throw exceptions only

Reply:
ok removed it